components/match2/commons/opponent.lua

```diff
@@ -0,0 +1,331 @@
+---
+-- @Liquipedia
+-- wiki=commons
+-- page=Module:Opponent
+--
+-- Please see https://github.com/Liquipedia/Lua-Modules to contribute
+--
+
+local Array = require('Module:Array')
+local Flags = require('Module:Flags')
+local String = require('Module:StringUtils')
+local Table = require('Module:Table')
+local TeamTemplate = require('Module:TeamTemplate')
+
+--[[
+Structural type representation of an opponent.
+
+Examples:
+{type = 'solo', players = {displayName = 'Neeb'}}
+{type = 'team', template = 'alpha x 2020'}
+{type = 'literal', name = 'B2'}
+
+See Opponent.isOpponent for the exact encoding scheme and required fields.
+
+- For opponent display components, see Module:OpponentDisplay.
+- For input from wikicode, use {{1Opponent|...}}, {{TeamOpponent|...}} etc.
+- Used by: PrizePool, GroupTableLeague, match2 Matchlist/Bracket,
+StarcraftMatchSummary
+- Wikis may add additional wiki-specific fields to the opponent representation.
+
+]]
+local Opponent = {}
+
+Opponent.partyTypes = {'solo', 'duo', 'trio', 'quad'}
+Opponent.types = Array.extend(Opponent.partyTypes, {'team', 'literal'})
+
+Opponent.partySizes = {
+	solo = 1,
+	duo = 2,
+	trio = 3,
+	quad = 4,
+}
+
+function Opponent.typeIsParty(type)
+	return Opponent.partySizes[type] ~= nil
+end
+
+--[[
+Returns the player count for a party type, or nil otherwise.
+
+Opponent.partySize('duo') == 2
+]]
+function Opponent.partySize(type)
+	return Opponent.partySizes[type]
+end
+
+--[[
+Creates a blank opponent of the specified type
+]]
+function Opponent.blank(type)
+	if type == 'team' then
+		return {type = type, template = 'tbd'}
+	elseif type == 'literal' then
+		return {type = type, name = ''}
+	else
```

Comment:
Explicitly add if? (everywhere else too)

Reply:
I don't think it's needed

Reply:
I do

Reply:
Why? Can you write out what you think it should look like

Reply:
When you're dealing with switch case type branching don't use `default` (or in this case `else`), it can cause types to go into this branch into the future even though we don't want them to. So, explicitly add ifs for every type instead.

Reply:
I have an assert in the else case. it will throw if it's not a party type in the else case.